fcli-core/fcli-tool/src/main/java/com/fortify/cli/tool/definitions/helper/ToolDefinitionsHelper.java

```diff
@@ -34,104 +38,264 @@
+		try {
+			UnirestHelper.download("tool", new URL(source).toString(), dest.toFile());
+		} catch (MalformedURLException e) {
+			if (!source.toLowerCase().endsWith(".zip") && !source.toLowerCase().endsWith(".yaml")) {
+				throw new FcliSimpleException("Invalid Tools definitions file");
+			}
+			mergeDefinitionsZip(dest);
+		}
+		return new ToolDefinitionsStateDescriptor(source, new Date(getModifiedTime(dest).toMillis()));
+	}
+
+	@SneakyThrows
+	private static boolean extractYamlFromZip(InputStream is, String yamlFile, Path outFile) {
+		try (ZipInputStream zis = new ZipInputStream(is)) {
+			ZipEntry entry;
+			while ((entry = zis.getNextEntry()) != null) {
+				if (entry.getName().equals(yamlFile)) {
+					Files.copy(zis, outFile, StandardCopyOption.REPLACE_EXISTING);
+					if (entry.getLastModifiedTime() != null) {
+						Files.setLastModifiedTime(outFile, entry.getLastModifiedTime());
+					}
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	@SneakyThrows
+	private static void mergeDefinitionsZip(Path dest) {
+		var yamlFileNames = new ArrayList<>(getRequiredYamlNames());
+		createDefinitionsStateDir(DEFINITIONS_STATE_DIR);
+		for (String yamlFile : yamlFileNames) {
+			boolean found = false;
+			Path outFile = DEFINITIONS_STATE_DIR.resolve(yamlFile);
+			if (!found && toolDefinitionCustomFilePath != null && Files.exists(Path.of(toolDefinitionCustomFilePath))) {
+				try (InputStream is = Files.newInputStream(Path.of(toolDefinitionCustomFilePath))) {
+					found = extractYamlFromZip(is, yamlFile, outFile);
+				}
+			}
+			if (!found && Files.exists(DEFINITIONS_STATE_ZIP)) {
+				try (InputStream is = Files.newInputStream(DEFINITIONS_STATE_ZIP)) {
+					found = extractYamlFromZip(is, yamlFile, outFile);
+				}
+			}
+			if (!found) {
+				try (InputStream is = FileUtils.getResourceInputStream(DEFINITIONS_INTERNAL_ZIP)) {
+					found = extractYamlFromZip(is, yamlFile, outFile);
+				}
+			}
+			if (!found) {
+				throw new FcliSimpleException("Required tool definition file missing: " + yamlFile);
+			}
+		}
+
+		FileTime latestTime = null;
+		try (java.util.zip.ZipOutputStream zos = new java.util.zip.ZipOutputStream(Files.newOutputStream(dest))) {
+			for (String yamlFile : yamlFileNames) {
+				Path filePath = DEFINITIONS_STATE_DIR.resolve(yamlFile);
+				if (Files.exists(filePath)) {
+					java.util.zip.ZipEntry zipEntry = new java.util.zip.ZipEntry(yamlFile);
+					FileTime fileTime = Files.getLastModifiedTime(filePath);
+					zipEntry.setLastModifiedTime(fileTime);
+					zos.putNextEntry(zipEntry);
+					Files.copy(filePath, zos);
+					zos.closeEntry();
+					if (latestTime == null || fileTime.compareTo(latestTime) > 0) {
+						latestTime = fileTime;
+					}
+				}
+			}
+		}
+		if (latestTime != null) {
+			Files.setLastModifiedTime(dest, latestTime);
+		}
+
+		for (String yamlFile : yamlFileNames) {
+			Path filePath = DEFINITIONS_STATE_DIR.resolve(yamlFile);
+			try {
+				Files.deleteIfExists(filePath);
```

Comment:
Apparently you're creating temporary files for each YAML file (and possibly also the downloaded zip file, I haven't checked); can you come up with a solution that doesn't require temporary files from being created and cleaned up again (as that's error-prone, for example files being left behind if update is interrupted by user or exception). Of course, we don't want to read the zip-file multiple times, so files may need to be fully processed in the order in which they appear in the zip-file, or you can use a zipfilesystem instead to access files in a zip file as if they were stored on the local file system.